Stephan/metrics count test

StephanDollberg opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.2.x
- [ ] v23.1.x
- [ ] v22.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/41507#018bf2dd-a7bc-492a-96e3-62a0ddfea20f:
```
"rptest.tests.metrics_test.MetricsTest.test_aggregate_metrics.aggregate_metrics=True"
```
<!--a32833e5-3ba5-4fa6-9244-7aa6948e9981-->

StephanDollberg commented:
/cdt
rp_version=pr
tests/rptest/tests/metrics_test.py::MetricsTest.test_new_metrics

StephanDollberg commented:
/cdt
rp_version=build
tests/rptest/tests/metrics_test.py::MetricsTest.test_new_metrics

StephanDollberg commented:
As discussed with @dotnwat, maybe produce some messages to make sure all lazy metrics have been instantiated.

StephanDollberg commented on tests/rptest/tests/metrics_test.py:
```diff
@@ -63,20 +133,128 @@ def test_aggregate_metrics(self, aggregate_metrics):
+        with open(golden_metrics_path(aggregate_metrics,
+                                      public_metrics)) as metrics_file:
+            golden_metrics = self.clean_metrics(
+                metrics_file.read().split("\n"))
```
done - switched to pathlib

StephanDollberg commented on tests/rptest/tests/metrics_test.py:
```diff
@@ -54,6 +121,9 @@ def test_aggregate_metrics(self, aggregate_metrics):
 
         """
+        def filter_metrics(metrics):
+            return list(metric for metric in metrics
```
done

StephanDollberg commented:
Rebased and:

 - Added a bit of produce and consume (with group)
 - Switched to single node - needed as the consumer group metrics are not replicated on all nodes
 - Applied minor things from above
 
> I think this only needs some thought whether it's worth reducing friction by removing metrics with low cardinality from the gold file. Up to you, but approved if you decide to use the current approach.

Need to think over this. I do like the friction as to make it explicit such that people "see" the cardinality they add but can see that it might become frustrating continuously having to update the file.

StephanDollberg commented:
> This is a big file with potentially lots of churn.

Can you clarify what you mean churn? Like lots of changes? Would that necessarily be a bad thing? One concern would be merge conflicts but given it's sorted I kinda expect it to be fine.

> I guess it's not so easy to only check metrics with labels of high cardinality, since they're mostly arbitrarily named.

We could certainly do something like this, for example ignore counters and gauges without `partition`, `topic` and `group` labels.

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/44952#018d8f87-0ca8-466d-845a-d726cbd28187
<!--1e100a69-6d6d-4eb1-ba94-3d3a5f77240f-->

BenPope reviewed (commented):
What's the current plan for this?

StephanDollberg commented:
> What's the current plan for this?

I wanted to merge it after the 24.1 merge but then was distracted with incidents as always so didn't get to rebasing it.

Also was kinda waiting for your thoughts on my reply above.
